pkg/tracker/reward/reward.go

```diff
@@ -0,0 +1,412 @@
+			<-ticker.C
+			continue
+		}
+		break
+	}
+
+	for {
+		select {
+		case <-self.ctx.Done():
+			return
+		case err := <-sub.Err():
+			if err != nil {
+				level.Error(logger).Log(
+					"msg",
+					"subscription error",
+					"err", err)
+			}
+
+			// Trying to resubscribe until it succeeds.
+			for {
+				select {
+				case <-self.ctx.Done():
+					return
+				default:
+				}
+				sub, err = self.nonceSubmittedSub(events)
+				if err != nil {
+					level.Error(logger).Log("msg", "re-subscribing to events failed", "err", err)
+					<-ticker.C
+					continue
+				}
+				break
+			}
+			level.Info(logger).Log("msg", "re-subscribed to events")
+		case event := <-events:
+			err := self.recordGasUsage(event)
+			if err != nil {
+				level.Error(self.logger).Log("msg", "record gas usage", "err", err)
+			}
+			err = self.recordGasUsageEstimated(self.ctx)
+			if err != nil {
+				level.Error(self.logger).Log("msg", "record gas usage estimation", "err", err)
+			}
+		}
+	}
+}
+
+func (self *RewardTracker) recordGasUsage(event *tellor.TellorNonceSubmitted) error {
+	receipt, err := self.client.TransactionReceipt(self.ctx, event.Raw.TxHash)
+	if err != nil {
+		return errors.Wrap(err, "receipt retrieval")
+	} else if receipt != nil && receipt.Status == types.ReceiptStatusSuccessful { // Failed transactions cost is monitored in a different process.
+		// tx, _, err := self.client.TransactionByHash(self.ctx, event.Raw.TxHash)
+		// if err != nil {
+		// 	level.Error(self.logger).Log("msg", "get transaction by hash", "err", err)
+		// 	return
+		// }
+		level.Debug(self.logger).Log("msg", "adding gas used", "gasUsed", receipt.GasUsed)
+		if err := self.addGasUsed(event.Slot.String(), receipt.GasUsed); err != nil {
+			return errors.Wrap(err, "adding gas used")
+		}
+		return nil
+	}
+	return errors.New("transaction not yet mined")
+}
+
+func (self *RewardTracker) addGasUsed(slot string, gasUsed uint64) error {
+	var err error
+	appender := self.tsDB.Appender(self.ctx)
+
+	// Round up the time so that all appends happen with the same TS and
+	// avoid out of order samples errors.
+	ts := timestamp.FromTime(time.Now().Round(5 * time.Second))
+
+	defer func() { // An appender always needs to be committed or rolled back.
+		if err != nil {
+			if err := appender.Rollback(); err != nil {
+				level.Error(self.logger).Log("msg", "db rollback failed", "err", err)
+			}
+			return
+		}
+		if errC := appender.Commit(); errC != nil {
+			err = errors.Wrap(err, "db append commit failed")
+		}
+	}()
+	lbls := labels.Labels{
+		labels.Label{Name: "__name__", Value: "gas_usage_actual"},
+		labels.Label{Name: "slot", Value: slot},
+	}
+
+	_, err = appender.Append(0, lbls, ts, float64(gasUsed))
+	if err != nil {
+		return errors.Wrap(err, "append values to the DB")
+	}
+	return nil
+}
+
+func (self *RewardTracker) recordGasUsageEstimated(ctx context.Context) error {
+	ctx, cncl := context.WithTimeout(ctx, 2*time.Second)
+	defer cncl()
+	// Getting abi.
```

Comment:
```suggestion
```
I think this and the other comments here are not necessary as the code is self explanatory